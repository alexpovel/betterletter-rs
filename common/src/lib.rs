use log::trace;

pub mod instrament;

pub fn titlecase(word: &str) -> String {
    let mut chars = word.chars();
    let mut result = String::with_capacity(word.len());

    if let Some(c) = chars.next() {
        for upper in c.to_uppercase() {
            result.push(upper);
        }
    }

    for c in chars {
        for lower in c.to_lowercase() {
            result.push(lower);
        }
    }

    result
}

pub fn is_compound_word(word: &str, predicate: &impl Fn(&str) -> bool) -> bool {
    trace!("Checking if word is valid compound word: '{}'", word);

    let indices = word.char_indices().skip(1);

    // Greedily fetch the longest possible prefix. Otherwise, we short-circuit and might
    // end up looking for (for example) "He" of "Heizölrechnung" and its suffix
    // "izölrechnung" (not a word), whereas we could have found "Heizöl" and "Rechnung"
    // instead.
    let mut highest_valid_index = None;
    for (i, _) in indices {
        let prefix = &word[..i];

        if predicate(prefix) {
            highest_valid_index = Some(i);
        }
    }

    match highest_valid_index {
        Some(i) => {
            let suffix = &word[i..];

            trace!(
                "Prefix '{}' found in word list, seeing if suffix '{}' is valid.",
                &word[..i],
                suffix
            );

            predicate(&titlecase(suffix))
                || predicate(suffix)
                || is_compound_word(&titlecase(suffix), predicate)
                || is_compound_word(suffix, predicate)
        }
        None => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case("hello", "Hello")]
    #[case("bItTe", "Bitte")]
    #[case("dANKE", "Danke")]
    #[case("übel", "Übel")]
    #[case("uebel", "Uebel")]
    #[case("😀", "😀")]
    #[case("ßuper", "SSuper")]
    #[case("ẞuperduper", "ẞuperduper")]
    #[case("WOW!!", "Wow!!")]
    #[case("ẞß", "ẞß")]
    fn test_titlecase(#[case] word: &str, #[case] expected: &str) {
        assert_eq!(titlecase(word), expected);
    }

    const WORDS: &[&str] = &["Süßwasser", "schwimm", "Bäder", "Mauer", "Dübel", "Kübel"];

    #[rstest]
    #[case("Süßwasserschwimmbäder", true)]
    #[case("Mauerdübel", true)]
    #[case("Mauerdübelkübel", true)]
    #[case("Not a compound word", false)]
    #[case("Mauer好", false)]
    #[case("Mauerdjieojoid", false)]
    fn test_is_compound_word(#[case] word: &str, #[case] expected: bool) {
        assert_eq!(is_compound_word(word, &|w| WORDS.contains(&w)), expected);
    }
}
