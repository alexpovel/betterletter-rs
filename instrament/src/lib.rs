pub fn sanitize_for_filename_use(filename: &str) -> String {
    const REPLACEMENT: char = '_';
    filename
        .replace(
            [
                ' ', ':', '<', '>', '\"', '/', '\\', '|', '?', '*', '\n', '\r',
            ],
            &REPLACEMENT.to_string(),
        )
        // Collapse consecutive underscores into one
        .split(REPLACEMENT)
        .filter(|&s| !s.is_empty())
        .collect::<Vec<_>>()
        .join(&REPLACEMENT.to_string())
}

#[macro_export]
macro_rules! instrament {
    ($(#[$attr:meta])* fn $name:ident ( $( $(#[$arg_attr:meta])* $arg:ident : $type:ty),* ) $body:expr ) => {
        ::paste::paste! {
            #[derive(::serde::Serialize)]
            struct [<$name:camel>]<'a> {
                $( $arg: &'a $type, )*
            }

            impl<'a> ::std::fmt::Display for [<$name:camel>]<'a> {
                fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
                    let items: Vec<String> = vec![
                        stringify!($name).to_owned() ,
                        $(
                            format!("{:#?}", self.$arg).to_owned() ,
                        )*
                    ];

                    let name = $crate::sanitize_for_filename_use(&items.join("-"));
                    write!(f, "{}", name)
                }
            }

            $(#[$attr])*
            fn $name ( $( $(#[$arg_attr])* $arg : $type),* ) {
                let function_data = [<$name:camel>] { $($arg: &$arg),* };
                let mut settings = ::insta::Settings::clone_current();
                settings.set_info(&function_data);

                settings.bind(|| {
                    #[allow(clippy::redundant_closure_call)]
                    $body(&function_data);
                });
            }
        }
    };
}
