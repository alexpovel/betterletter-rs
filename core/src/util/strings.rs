#[derive(Debug, PartialEq, Eq)]
pub(crate) enum WordCasing {
    AllLowercase,
    AllUppercase,
    Titlecase,
    Mixed,
}

impl TryFrom<&str> for WordCasing {
    type Error = &'static str;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        if value.is_empty() {
            return Err("String is empty");
        }

        let mut has_lowercase = false;
        let mut has_uppercase = false;
        let mut is_titlecase = true;

        for (i, c) in value.chars().enumerate() {
            if c.is_lowercase() {
                has_lowercase = true;

                if i == 0 {
                    is_titlecase = false;
                }
            } else if c.is_uppercase() {
                has_uppercase = true;

                if i != 0 {
                    is_titlecase = false;
                }
            } else {
                return Err("String contains characters with undecidable casing");
            }
        }

        match (is_titlecase, has_lowercase, has_uppercase) {
            (true, _, _) => Ok(Self::Titlecase),
            (_, true, false) => Ok(Self::AllLowercase),
            (_, false, true) => Ok(Self::AllUppercase),
            (_, true, true) => Ok(Self::Mixed),
            (_, false, false) => unreachable!("Impossible case: any non-empty string has either lower- or uppercase or returned an `Err` early."),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use common::instrament;
    use rstest::rstest;
    use serde::Serialize;

    impl Serialize for WordCasing {
        fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
            match self {
                Self::AllLowercase => serializer.serialize_str("AllLowercase"),
                Self::AllUppercase => serializer.serialize_str("AllUppercase"),
                Self::Titlecase => serializer.serialize_str("Titlecase"),
                Self::Mixed => serializer.serialize_str("Mixed"),
            }
        }
    }

    instrament! {
            #[rstest]
            fn test_word_casing_from_string(
            #[values(
                "hello",
                "bItTe",
                "dANKE",
                "Ã¼bel",
                "uebel",
                "ðŸ˜€",
                "ÃŸuper",
                "áºžuperduper",
                "WOW!!",
                "SCREAMING",
                "áºžÃŸ",
                "",
            )]
                word: String
            ) (|data: &TestWordCasingFromString| {
                insta::assert_yaml_snapshot!(data.to_string(), WordCasing::try_from(word.as_str()));
            }
        )
    }
}
