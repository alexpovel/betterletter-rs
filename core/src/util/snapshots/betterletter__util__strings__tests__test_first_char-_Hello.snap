---
source: core/src/util/strings.rs
expression: first_char(&word).to_string()
info:
  word: Hello
---
H
